presto-raptorx/src/main/java/com/facebook/presto/raptorx/systemtable/TableSystemTable.java

```diff
@@ -0,0 +1,174 @@
+import static com.facebook.presto.raptorx.util.PredicateUtil.listTables;
+import static com.facebook.presto.spi.SystemTable.Distribution.SINGLE_COORDINATOR;
+import static com.facebook.presto.spi.predicate.TupleDomain.extractFixedValues;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class TableSystemTable
+        implements SystemTable
+{
+    private final TransactionManager transactionManager;
+    private final ConnectorTableMetadata tableMetadata;
+
+    @Inject
+    public TableSystemTable(TransactionManager transactionManager, TypeManager typeManager)
+    {
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+
+        Type varcharArray = typeManager.getType(parseTypeSignature("array<varchar>"));
+        this.tableMetadata = new ConnectorTableMetadata(
+                new SchemaTableName("system", "tables"),
+                ImmutableList.<ColumnMetadata>builder()
+                        .add(new ColumnMetadata("table_schema", VARCHAR))
+                        .add(new ColumnMetadata("table_name", VARCHAR))
+                        .add(new ColumnMetadata("temporal_column", VARCHAR))
+                        .add(new ColumnMetadata("ordering_columns", varcharArray))
+                        .add(new ColumnMetadata("distribution_name", VARCHAR))
+                        .add(new ColumnMetadata("bucket_count", BIGINT))
+                        .add(new ColumnMetadata("bucketing_columns", varcharArray))
+                        .build());
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return tableMetadata;
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        Transaction transaction = transactionManager.get(transactionHandle);
+
+        Map<Integer, NullableValue> values = extractFixedValues(constraint).orElse(ImmutableMap.of());
+        Optional<String> schemaName = getStringValue(values.get(getColumnIndex(tableMetadata, "table_schema")));
+        Optional<String> tableName = getStringValue(values.get(getColumnIndex(tableMetadata, "table_name")));
+
+        // TODO: fetch table info in bulk
+        List<TableInfo> tables = listTables(transaction, schemaName, tableName).stream()
+                .map(transaction::getTableId)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .map(transaction::getTableInfo)
+                .collect(toImmutableList());
```

Comment:
"highker on Jan 30  Member
Maybe put this code into the following loop to avoid materialize internal list?"